gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java

```diff
@@ -156,33 +174,136 @@ public void handleResponse(WriteObjectResponse response) {
             new VerificationAttributes(md5Hash, crc32c));
   }
 
-  @Override
-  public void startUpload(InputStream pipeSource) {
-    // Given that the two ends of the pipe must operate asynchronous relative
-    // to each other, we need to start the upload operation on a separate thread.
+  public boolean isOpen() {
+    return (writeBuffer != null);
+  }
+
+  public synchronized int write(ByteBuffer buffer) throws IOException {
+    checkState(initialized, "initialize() must be invoked before use.");
+    int bytesWritten = 0;
+    if (!isOpen()) {
+      throw new ClosedChannelException();
+    }
+
+    // No point in writing further if upload failed on another thread.
+    if (uploadOperation.isDone()) {
+      waitForCompletionAndThrowIfUploadFailed();
+    }
+
+    // create chunks of MAX_BYTES_PER_MESSAGE and push them into writeBuffer. If the loop ends up
+    // in a scenario where the lastChunk size is less than MAX_BYTES_PER_MESSAGE, then we hold on
+    // to it until next write or close.
+    while (buffer.remaining() > 0) {
+      int bytesToWrite = 0;
+      if (lastChunk != null) {
+        bytesToWrite = min(buffer.remaining(), MAX_BYTES_PER_MESSAGE - lastChunk.size());
+        lastChunk = lastChunk.concat(ByteString.copyFrom(buffer, bytesToWrite));
+      } else {
+        bytesToWrite = min(buffer.remaining(), MAX_BYTES_PER_MESSAGE);
+        lastChunk = ByteString.copyFrom(buffer, bytesToWrite);
```

Comment:
This is actually doing a copy only - could you please check.

Reply:
confirmed that this is fine. Earlier there were 2 memcpy. With this implementation now there is only one.